QtAcademyProject.cpp

```diff
@@ -12,22 +12,155 @@
+  }
+}
+
+void QtAcademyProject::loadOfflineBasemaps()
+{
+  // Check to see if any offline data exists
+  if (QFile::exists(vtpkPath+"/vectorTiles.vtpk"))
+  {
+    // Create a VectorTileCache object from the offline data
+    m_vectorTileCache = new VectorTileCache(vtpkPath + "/vectorTiles.vtpk", m_tempObject.get());
+
+    // Check to see if any additional item resources exist and use them in the constructor if so
+    if (QFile::exists(vtpkPath+"/itemResources"))
+    {
+      m_itemResourceCache = new ItemResourceCache(vtpkPath+"/itemResources", m_tempObject.get());
+      m_offlineLayer = new ArcGISVectorTiledLayer(m_vectorTileCache, m_itemResourceCache, m_tempObject.get());
+    }
+    else
+    {
+      m_offlineLayer = new ArcGISVectorTiledLayer(m_vectorTileCache, m_tempObject.get());
+    }
+
+    // Add the created ArcGISVectorTileLayer to the base layers
+    m_map->basemap()->baseLayers()->append(m_offlineLayer);
+  }
+}
+
+void QtAcademyProject::createOfflineAreaFromExtent()
+{
+  // Export all available ArcGISVectorTileLayers
+  for (Layer* layer : *(m_basemap->baseLayers()))
+  {
+    if (ArcGISVectorTiledLayer* vectorTileLayer = dynamic_cast<ArcGISVectorTiledLayer*>(layer))
+      exportVectorTiles(vectorTileLayer);
+  }
+}
+
+void QtAcademyProject::exportVectorTiles(ArcGISVectorTiledLayer* vectorTileLayer)
+{
+  // Create a new export task from the layer's source url
+  m_exportVectorTilesTask = new ExportVectorTilesTask(vectorTileLayer->url(), m_tempObject.get());
+
+  // Create default parameters for the layer service
+  // Normalize the central meridian in case the download area crosses the meridian
+  m_exportVectorTilesTask->createDefaultExportVectorTilesParametersAsync(GeometryEngine::normalizeCentralMeridian(m_mapView->visibleArea()), m_mapView->mapScale()*0.1)
+      .then(this, [this](ExportVectorTilesParameters defaultParams)
+  {
+    QDir path(vtpkPath);
+
+    // Remove any existing offline files
+    path.removeRecursively();
+
+    // Create the directory if it does not already exist
+    if (!QDir().exists(vtpkPath))
+      QDir().mkdir(vtpkPath);
+
+    const QString vtpkFileName = vtpkPath+"/vectorTiles.vtpk";
+    const QString itemResourcesPath = vtpkPath+"/itemResources";
+
+    // Create a Job object to manage the export
+    m_exportVectorTilesJob = m_exportVectorTilesTask->exportVectorTiles(defaultParams, vtpkFileName, itemResourcesPath);
+
+    // Monitor the download progress and update the UI every time it changes
+    connect(m_exportVectorTilesJob, &Job::progressChanged, this, [this]()
+    {
+      m_downloadProgress = m_exportVectorTilesJob->progress();
+      emit downloadProgressChanged();
+    });
+
+    // Display any errors in the console log
+    connect(m_exportVectorTilesJob, &Job::errorOccurred, this, [](const Error& e){qWarning() << e.message() << e.additionalMessage();});
+
+    // Once all async slots have been created, start the export
+    m_exportVectorTilesJob->start();
+  });
```

Comment:
For completeness, I recommend adding an onFailed in case anything goes wrong. It shows best practices as well as our ability to report errors.